### PR TITLE
introduce new `norm` operator that can differentiate through `lambda x: norm(x-x)` + lighter tests across the board

### PR DESCRIPTION
**Pull request by marcocuturi**
Computing the gradient of Sinkhorn divergences using the `Euclidean` cost results in `NaN` values, because of the reliance on `jnp.linalg.norm`, and in particular, differentiation of the distance of a point against itself, e.g. `norm(x-x)` in the diagonal of a symmetric cost matrix. 

The fact that such a gradient is (and should be, in general) `NaN` is well documented, see e.g. https://github.com/google/jax/issues/6484 

However, In the context of OT, this poses problems, since it is safe to ignore these contributions, and therefore treat them as having 0 gradient.

This PR introduces a new `norm` function that does not blow up with a `NaN` at `0`.

```
f = lambda x: ott.math.utils.norm(x - x)
g = lambda x: jnp.linalg.norm(x - x)
x = jnp.array([1.2, 3.2, 4.1])
print(jax.grad(f)(x))
print(jax.grad(g)(x))
```

results in 

```
[0. 0. 0.]
[nan nan nan]
```

as a result it should be now possible to differentiate through a `sinkhorn_divergence` with a more elaborate cost without producing `NaN`'s.

Also, in order to speed things up in CI, prune out some tests.

### CODE REVIEW

**Comment by marcocuturi** [comment]
also of interest to @theouscidda6 as potentially useful in Monge gap.

File: src/ott/math/utils.py [hunk 1]
```diff
@@ -37,6 +43,41 @@ def safe_log(  # noqa: D103
   return jnp.where(x > 0., jnp.log(x), jnp.log(eps))
 
 
+@functools.partial(jax.custom_jvp, nondiff_argnums=[1])
+def norm(x, ord=2.0):
```
Comment (michalk8) [inline src/ott/math/utils.py:47]: I think this function should accept all arguments of `jnp.linalg.norm` - `axis` and `keepdims` as well.

Comment (michalk8) [reply src/ott/math/utils.py:47]: Please also add type hints.

File: tests/math/math_utils_test.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import jax
+import jax.numpy as jnp
+from ott.math import utils as mu
+
+
+class TestNorm:
+
+  def test_norm(self, rng: jax.random.PRNGKeyArray):
```
Comment (michalk8) [inline tests/math/math_utils_test.py:21]: Please parameterize by at least `ord` (later would also check for `axis`).

File: tests/math/math_utils_test.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    d = 13
+    f = lambda x: mu.norm(x - x)
+    g = lambda x: jnp.linalg.norm(x - x)
+    x = jax.random.uniform(rng, (d,))
+    assert jnp.all(jax.grad(f)(x) == 0.0)
+    assert jnp.all(jnp.isnan(jax.grad(g)(x)))
```
Comment (michalk8) [inline tests/math/math_utils_test.py:27]: Same as above, would use `np.testing.assert_array_equal(..., np.nan)`

File: tests/math/math_utils_test.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import jax
+import jax.numpy as jnp
+from ott.math import utils as mu
+
+
+class TestNorm:
+
```
Comment (michalk8) [inline tests/math/math_utils_test.py:20]: Please add a new parameterized that checks the new `mu.norm` is the same for non-0 input as `jnp.linalg.norm`.

File: tests/tools/sinkhorn_divergence_test.py [hunk 5]
```diff
@@ -70,15 +71,19 @@ def test_euclidean_point_cloud(self, cost_fn: costs.CostFn, epsilon: float):
     div2 -= 0.5 * sinkhorn.solve(geometry_xx, self._a, self._a).reg_ot_cost
     div2 -= 0.5 * sinkhorn.solve(geometry_yy, self._b, self._b).reg_ot_cost
 
-    np.testing.assert_allclose(div.divergence, div2, rtol=1e-5, atol=1e-5)
+    np.testing.assert_allclose(out.divergence, div2, rtol=1e-5, atol=1e-5)
 
-    # Test div of x to itself close to 0.
+    # Check differentiability of Sinkhorn divergence works, without NaN's.
+    grad = jax.grad(lambda x: div(x).divergence)(x)
```
Comment (michalk8) [inline tests/tools/sinkhorn_divergence_test.py:77]: I'd consider jitting this.

File: tests/tools/sinkhorn_divergence_test.py [hunk 6]
```diff
@@ -70,15 +71,19 @@ def test_euclidean_point_cloud(self, cost_fn: costs.CostFn, epsilon: float):
...
+    assert jnp.all(jnp.logical_not(jnp.isnan(grad)))
```
Comment (michalk8) [inline tests/tools/sinkhorn_divergence_test.py:78]: Use `np.testing...`.

File: tests/math/math_utils_test.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  def test_norm(self, rng: jax.random.PRNGKeyArray):
+    d = 13
+    f = lambda x: mu.norm(x - x)
+    g = lambda x: jnp.linalg.norm(x - x)
+    x = jax.random.uniform(rng, (d,))
+    assert jnp.all(jax.grad(f)(x) == 0.0)
```
Comment (michalk8) [inline tests/math/math_utils_test.py:26]: Please also add a test for `jax.hessian`. [UPDATE] Running:
```python
import jax
import jax.numpy as jnp
from ott.math import utils as mu

rng = jax.random.PRNGKey(0)

d = 100
f = lambda x: mu.norm(x)
g = lambda x: jnp.linalg.norm(x)
x = jax.random.uniform(rng, (d,))

h_g = jax.hessian(g)(x)
h_f = jax.hessian(f)(x)

jnp.max(jnp.abs(h_f - h_g))  # Array(0.16832732, dtype=float32)
```
doesn't seem to hold. I guess this is to be expected, if we define the custom JVP on the `norm`; maybe having a custom JVP for `def h(z: jnp.ndarray): ...` would solve this?

File: src/ott/math/utils.py [hunk 8]
```diff
@@ -37,6 +43,41 @@ def safe_log(  # noqa: D103
...
+  contribute to an OT cost (when, for instance, computing Sinkhorn divergences,
+  involving computing the OT cost of a point cloud with itself).
+
+  To avoid such `NaN` values, this custom norm implementation uses the
+  double-where trick, to avoid having branches that output any `NaN`, and
+  safely output a 0 instead.
+  """
+  x, = primals
+  x_is_zero = jnp.all(jnp.logical_not(x))
+  clean_x = jnp.where(x_is_zero, jnp.ones(x.shape), x)
```
Comment (michalk8) [inline src/ott/math/utils.py:74]: Use `jnp.ones_like(x)`.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: tests/math/math_utils_test.py (hunk 4 above)
Comment (michalk8) [reply tests/math/math_utils_test.py:20]: This might be helpful (to compare with finite difference as well; we should use it in other places): `jax.test_util.check_grads`.

File: src/ott/math/utils.py (hunk 1 above)
Comment (marcocuturi) [reply src/ott/math/utils.py:47]: Not completey sure we need the `axis` and `keepdims`, these tend to help mostly when computing matrix norms, which haven't appeared yet in our use-case as a ground cost. But will add, as long as this does not cause confusions or difficulties!

File: src/ott/math/utils.py [hunk 9]
```diff
@@ -37,6 +43,41 @@ def safe_log(  # noqa: D103
...
+  such distances between a point and itself can be safely ignored when they
+  contribute to an OT cost (when, for instance, computing Sinkhorn divergences,
+  involving computing the OT cost of a point cloud with itself).
+
+  To avoid such `NaN` values, this custom norm implementation uses the
+  double-where trick, to avoid having branches that output any `NaN`, and
+  safely output a 0 instead.
+  """
+  x, = primals
+  x_is_zero = jnp.all(jnp.logical_not(x))
```
[the remark this answers is not in the document]
Comment (marcocuturi) [reply src/ott/math/utils.py:73]: i think this is correct? if this returns `True`, then all elements of `jnp.logical_not(x)` are true. An element of `jnp.logical_not(x)` is True only if `x` is `False`, i.e. `x` is 0.

File: tests/math/math_utils_test.py (hunk 4 above)
Comment (marcocuturi) [reply tests/math/math_utils_test.py:20]: thanks! there was indeed a bug (inverted args in `jnp.where`)

File: tests/tools/sinkhorn_divergence_test.py (hunk 6 above)
Comment (marcocuturi) [reply tests/tools/sinkhorn_divergence_test.py:78]: not sure I can find the adequate test in there, https://numpy.org/doc/stable/reference/routines.testing.html

File: tests/math/math_utils_test.py (hunk 7 above)
Comment (marcocuturi) [reply tests/math/math_utils_test.py:26]: probably due to my bug before! now passes

**Comment by marcocuturi** [comment]
thanks Michal!

File: src/ott/math/utils.py [hunk 10]
```diff
@@ -37,6 +43,70 @@ def safe_log(  # noqa: D103
   return jnp.where(x > 0., jnp.log(x), jnp.log(eps))
 
 
+@functools.partial(jax.custom_jvp, nondiff_argnums=[1, 2, 3])
```
Comment (michalk8) [inline src/ott/math/utils.py:46]: To reduce this, I think we can do:
```python
norm = jax.custom_jvp(jax.jit(jnp.linalg.norm, static_argnames=("ord", "axis", "keepdims"), nondiff_argnums=[1, 2, 3])
```

File: src/ott/math/utils.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (michalk8) [reply src/ott/math/utils.py:101]: You're right, sorry for the mistake.

File: tests/tools/sinkhorn_divergence_test.py (hunk 6 above)
Comment (michalk8) [reply tests/tools/sinkhorn_divergence_test.py:80]: You can use `np.testing.assert_array_equal(jnp.isnan(grad), False)`.

File: tests/math/math_utils_test.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    ds = (4, 7, 3)
+    z = jax.random.uniform(rng, ds)
+    h = lambda x: mu.norm(x - x, ord=ord, axis=1)
+    jac = jax.jacobian(h)(z)
+    np.testing.assert_array_equal(jac, 0.0)
+
+    # Check native jax's norm still returns NaNs when ord is float.
+    if ord != jnp.inf:
+      g = lambda x: jnp.linalg.norm(x - x, ord=ord)
+      assert jnp.all(jnp.isnan(jax.grad(g)(x)))
```
Comment (michalk8) [inline tests/math/math_utils_test.py:48]: Here I'd use `np.testing.assert_array_equal(jnp.isnan(jax.grad(g)(x)), True)`.

File: tests/math/math_utils_test.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      h = lambda x: jnp.linalg.norm(x - x, ord=ord, axis=1)
+      jac = jax.jacobian(h)(z)
+      assert jnp.all(jnp.isnan(jac))
```
Comment (michalk8) [inline tests/math/math_utils_test.py:51]: Would use `np.testing.assert_array_equal(jnp.isnan(jac), True)`.

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: src/ott/math/utils.py (hunk 10 above)
Comment (marcocuturi) [reply src/ott/math/utils.py:46]: as discussed, will keep this, because of flexibility to have a custom pydocs.
